File: folkmq/src/main/java/org/noear/folkmq/client/MqRouter.java
package org.noear.folkmq.client;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

/**
 * 消息路由器
 *
 * @author noear
 * @since 1.3
 */
public class MqRouter implements MqConsumeHandler {
    private final Function<MqMessageReceived, String> mappingHandler;
    private Map<String, MqConsumeHandler> mappingMap = new HashMap<>();
    private MqConsumeHandler consumeHandler;

    public MqRouter(Function<MqMessageReceived, String> mappingHandler) {
        this.mappingHandler = mappingHandler;
    }

    public MqRouter doOn(String mapping, MqConsumeHandler consumeHandler) {
        mappingMap.put(mapping, consumeHandler);
        return this;
    }

    public MqRouter doOnConsume(MqConsumeHandler consumeHandler) {
        this.consumeHandler = consumeHandler;
        return this;
    }

    @Override
    public void consume(MqMessageReceived message) throws Exception {
        if (consumeHandler != null) {
            consumeHandler.consume(message);
        }

        String mapping = mappingHandler.apply(message);
        MqConsumeHandler handler = mappingMap.get(mapping);
        if (handler != null) {
            handler.consume(message);
        }
    }
}

File: folkmq-lang-python/folkmq/folkmq/client/MqRouter.py
from typing import Union, Callable, Dict

from folkmq.client.MqMessageReceived import MqMessageReceived


# 消息路由器
class MqRouter:
    def __init__(self, mappingHandler: Callable[[MqMessageReceived], str]):
        self._mappingHandler:Callable[[MqMessageReceived], str] = mappingHandler
        self._mappingMap:Dict[str,str] = {}
        self._consumeHandler:Union[Callable[[MqMessageReceived], None], None] = None

    # 添加映射处理
    def do_on(self, mapping:str, consumeHandler:Callable[[MqMessageReceived], None]) -> 'MqRouter':
        self._mappingMap[mapping] = consumeHandler
        return self

    # 添加消费处理
    def do_on_consume(self, consumeHandler:Callable[[MqMessageReceived], None]) -> 'MqRouter':
        self._consumeHandler = consumeHandler
        return self

    # 消费
    def consume(self, message:MqMessageReceived):
        if self._consumeHandler:
            self._consumeHandler(message)

        mapping = self._mappingHandler(message)
        handler = self._mappingMap.get(mapping)
        if handler:
            handler(message)